README.md
nginx Cookbook
==============

Installs nginx from package and sets up configuration handling.

Requirements
------------
### Cookbooks
The following cookbooks are direct dependencies because they're used for common "default" functionality.

- ohai (for nginx::ohai_plugin)

The following cookbook is not a strict dependency because its use can be controlled by an attribute, so it may not be a common "default."

- When using Nginx.org's stable package, `recipe[apt::default]` is required.


### Platforms
The following platforms are supported and tested under test kitchen:

- Ubuntu 14.04

Other Debian family distributions are assumed to work.


Attributes
----------
Node attributes for this cookbook are logically separated into different files. Some attributes are set only via a specific recipe.

### default
Generally used attributes. Some have platform specific values. See `attributes/default.rb`. "The Config" refers to "nginx.conf" the main config file.

## default.rb

Generally used attributes. Some have platform specific values. See
`attributes/default.rb`. "The Config" refers to "nginx.conf" the main
config file.

**v0.101.0 - Attribute Change**: `node['nginx']['url']` is now
  `node['nginx']['source']['url']` as the URL was only used when
  retrieving the source to build Nginx.

- `node['nginx']['dir']` - Location for Nginx configuration.
- `node['nginx']['conf_template']` - The `source` template to use when creating the `nginx.conf`.
- `node['nginx']['conf_cookbook']` - The cookbook where `node['nginx']['conf_template']` resides.
- `node['nginx']['log_dir']` - Location for Nginx logs.
- `node['nginx']['log_dir_perm']` - Permissions for Nginx logs folder.
- `node['nginx']['user']` - User that Nginx will run as.
- `node['nginx']['group]` - Group for Nginx.
- `node['nginx']['port']` - Port for nginx to listen on.
- `node['nginx']['binary']` - Path to the Nginx binary.
- `node['nginx']['default_hostname']` - The `server_name` for the `default` server.
- `node['nginx']['default_return_code']` - If set, requests to `default_hostname` will return this instead of trying to load something. Example values are `404` or `301 https://$server_name$request_uri`.
- `node['nginx']['upstart']['foreground']` - Set this to true if you
  want upstart to run nginx in the foreground, set to false if you
  want upstart to detach and track the process via pid.
- `node['nginx']['upstart']['runlevels']` - String of runlevels in the
  format '2345' which determines which runlevels nginx will start at
  when entering and stop at when leaving.
- `node['nginx']['upstart']['respawn_limit']` - Respawn limit in upstart
  stanza format, count followed by space followed by interval in seconds.
- `node['nginx']['pid']` - Location of the PID file.
- `node['nginx']['keepalive']` - Whether to use `keepalive_timeout`,
  any value besides "on" will leave that option out of the config.
- `node['nginx']['keepalive_requests']` - used for config value of
  `keepalive_requests`.
- `node['nginx']['keepalive_timeout']` - used for config value of
  `keepalive_timeout`.
- `node['nginx']['worker_processes']` - used for config value of
  `worker_processes`.
- `node['nginx']['worker_connections']` - used for config value of
  `events { worker_connections }`
- `node['nginx']['worker_rlimit_nofile']` - used for config value of
  `worker_rlimit_nofile`. Can replace any "ulimit -n" command. The
  value depend on your usage (cache or not) but must always be
  superior than worker_connections.
- `node['nginx']['multi_accept']` - used for config value of `events {
  multi_accept }`. Try to accept() as many connections as possible.
  Disable by default.
- `node['nginx']['event']` - used for config value of `events { use
  }`. Set the event-model. By default nginx looks for the most
  suitable method for your OS.
- `node['nginx']['accept_mutex_delay']` - used for config value of
  `accept_mutex_delay`
- `node['nginx']['server_tokens']` - used for config value of
  `server_tokens`.
- `node['nginx']['server_names_hash_bucket_size']` - used for config
  value of `server_names_hash_bucket_size`.
- `node['nginx']['disable_access_log']` - set to true to disable the
  general access log, may be useful on high traffic sites.
- `node['nginx']['access_log_options']` - Set to a string of additional options
  to be appended to the access log directive
- `node['nginx']['error_log_options']` - Set to a string of additional options
  to be appended to the error log directive
- `node['nginx']['default_site_enabled']` - enable the default site
- `node['nginx']['sendfile']` - Whether to use `sendfile`. Defaults to "on".
- `node['nginx']['tcp_nopush']` - Whether to use `tcp_nopush`. Defaults to "on".
- `node['nginx']['tcp_nodelay']` - Whether to use `tcp_nodelay`. Defaults to "on".
- `node['nginx']['types_hash_max_size']` - Used for the
  `types_hash_max_size` configuration directive.
- `node['nginx']['types_hash_bucket_size']` - Used for the
  `types_hash_bucket_size` configuration directive.
- `node['nginx']['proxy_read_timeout']` - defines a timeout (between two
  successive read operations) for reading a response from the proxied server.
- `node['nginx']['client_body_buffer_size']` - used for config value of
  `client_body_buffer_size`.
- `node['nginx']['client_max_body_size']` - specifies the maximum accepted body
  size of a client request, as indicated by the request header Content-Length.
* `node['nginx']['sts_max_age']` - Enable Strict Transport Security for all apps (See: http://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security).  This attribute adds the following header:

  Strict-Transport-Security max-age=SECONDS

to all incoming requests and takes an integer (in seconds) as its argument.
* `node['nginx']['default']['modules']` - Array specifying which
modules to enable via the conf-enabled config include function.
Currently the only valid value is "socketproxy".

Other configurations

- `node['nginx']['extra_configs']` - a Hash of key/values to nginx configuration.

Rate Limiting

- `node['nginx']['enable_rate_limiting']` - set to true to enable rate
  limiting (`limit_req_zone` in nginx.conf)
- `node['nginx']['rate_limiting_zone_name']` - sets the zone in
  `limit_req_zone`.
- `node['nginx']['rate_limiting_backoff']` - sets the backoff time for
  `limit_req_zone`.
- `node['nginx']['rate_limit']` - set the rate limit amount for
  `limit_req_zone`.

### gzip module

- `node['nginx']['gzip']` - Whether to use gzip, can be "on" or "off"
- `node['nginx']['gzip_http_version']` - used for config value of `gzip_http_version`.
- `node['nginx']['gzip_comp_level']` - used for config value of `gzip_comp_level`.
- `node['nginx']['gzip_proxied']` - used for config value of `gzip_proxied`.
- `node['nginx']['gzip_vary']` - used for config value of `gzip_vary`.
- `node['nginx']['gzip_buffers']` - used for config value of `gzip_buffers`.
- `node['nginx']['gzip_types']` - used for config value of `gzip_types` - must be an Array.
- `node['nginx']['gzip_min_length']` - used for config value of `gzip_min_length`.
- `node['nginx']['gzip_disable']` - used for config value of `gzip_disable`.
- `node['nginx']['gzip_static']` - used for config value of `gzip_static` (`http_gzip_static_module` must be enabled)
### Attributes set in recipes

#### nginx::authorized_ips
- `node['nginx']['remote_ip_var']` - The remote ip variable name to
  use.
- `node['nginx']['authorized_ips']` - IPs authorized by the module

#### nginx::http_realip_module
From: http://nginx.org/en/docs/http/ngx_http_realip_module.html

- `node['nginx']['realip']['header']` - Header to use for the RealIp
  Module; only accepts "X-Forwarded-For" or "X-Real-IP"
- `node['nginx']['realip']['addresses']` - Addresses to use for the
  `http_realip` configuration.
- `node['nginx']['realip']['real_ip_recursive']` - If recursive search is enabled, the original client address that matches one of the trusted addresses is replaced by the last non-trusted address sent in the request header field. Can be on "on" or "off" (default).

### geoip
These attributes are used in the `nginx::http_geoip_module` recipe.
Please note that the `country_dat_checksum` and `city_dat_checksum`
are based on downloads from a datacenter in Fremont, CA, USA. You
really should override these with checksums for the geo tarballs from
your node location.

**Note** The upstream, maxmind.com, may block access for repeated
  downloads of the data files. It is recommended that you download and
  host the data files, and change the URLs in the attributes.

- `node['nginx']['geoip']['enable_city']` - Whether to enable City
  data
- `node['nginx']['geoip']['country_dat_url']` - Country data tarball
  URL
- `node['nginx']['geoip']['country_dat_checksum']` - Country data
  tarball checksum
- `node['nginx']['geoip']['city_dat_url']` - City data tarball URL
- `node['nginx']['geoip']['city_dat_checksum']` - City data tarball
  checksum

### upload_progress
These attributes are used in the `nginx::upload_progress_module`
recipe.

- `node['nginx']['upload_progress']['url']` - URL for the tarball.
- `node['nginx']['upload_progress']['checksum']` - Checksum of the
  tarball.
- `node['nginx']['upload_progress']['javascript_output']` - Output in javascript.
  Default is `true` for backwards compatibility.
- `node['nginx']['upload_progress']['zone_name']` - Zone name which will
  be used to store the per-connection tracking information.
  Default is `proxied`.
- `node['nginx']['upload_progress']['zone_size']` - Zone size in bytes.
  Default is `1m` (1 megabyte).

### passenger
These attributes are used in the `nginx::passenger` recipe.

- `node['nginx']['passenger']['version']` - passenger gem version
- `node['nginx']['passenger']['root']` - passenger gem root path
- `node['nginx']['passenger']['install_rake']` - set to false if rake already present on system
- `node['nginx']['passenger']['max_pool_size']` - maximum passenger
  pool size (default=10)
- `node['nginx']['passenger']['ruby']` - Ruby path for Passenger to
  use (default=`$(which ruby)`)
- `node['nginx']['passenger']['spawn_method']` - passenger spawn
  method to use (default=`smart-lv2`)
- `node['nginx']['passenger']['buffer_response']` - turns on or off
  response buffering (default=`on`)
- `node['nginx']['passenger']['max_pool_size']` - passenger maximum
  pool size (default=`6`)
- `node['nginx']['passenger']['min_instances']` - minimum instances
  (default=`1`)
- `node['nginx']['passenger']['max_instances_per_app']` - maximum
  instances per app (default=`0`)
- `node['nginx']['passenger']['pool_idle_time']` - passenger pool idle
  time (default=`300`)
- `node['nginx']['passenger']['max_requests']` - maximum requests
  (default=`0`)
- `node['nginx']['passenger']['nodejs']` - Nodejs path for Passenger to
  use (default=nil)

Basic configuration to use the official Phusion Passenger repositories:
- `node['nginx']['package_name']` - 'nginx-extras'
- `node['nginx']['passenger']['install_method']` - 'package'

### status
These attributes are used in the `nginx::http_stub_status_module` recipe.

- `node['nginx']['status']['port']` - The port on which nginx will
  serve the status info (default: 8090)

## socketproxy.rb

These attributes are used in the `nginx::socketproxy` recipe.

* `node['nginx']['socketproxy']['root']` - The directory (on your server) where socketproxy apps are deployed.
* `node['nginx']['socketproxy']['default_app']` - Static assets directory for requests to "/" that don't meet any proxy_pass filter requirements.
* `node['nginx']['socketproxy']['apps']['app_name']['prepend_slash']` - Prepend a slash to requests to app "app_name" before sending them to the socketproxy socket.
* `node['nginx']['socketproxy']['apps']['app_name']['context_name']` - URI (e.g. "app_name" in order to achieve "http://mydomain.com/app_name") at which to host the application "app_name"
* `node['nginx']['socketproxy']['apps']['app_name']['subdir']` - Directory (under `node['nginx']['socketproxy']['root']`) in which to find the application.

Recipes
-------
This cookbook provides three main recipes for installing Nginx.

- `default.rb` - *Use this recipe* if you have a native package for
  Nginx.
- `repo.rb` - The developer of Nginx also maintain
  [stable packages](http://nginx.org/en/download.html) for several
  platforms.

### default
The default recipe will install Nginx as a native package for the
system through the package manager and sets up the configuration
according to the Debian site enable/disable style with `sites-enabled`
using the `nxensite` and `nxdissite` scripts. The nginx service will
be managed with the normal init scripts that are presumably included
in the native package.

Includes the `ohai_plugin` recipe so the plugin is available.

### socketproxy

This will add socketproxy support to your nginx proxy setup.  Do not
include this recipe directly.  Instead, add it to the
`node['nginx']['default']['modules']` array (see below).

### ohai_plugin

This recipe provides an Ohai plugin as a template. It is included by the `default` recipe.

### authorized_ips
Sets up configuration for the `authorized_ip` nginx module.

Definitions
-----------

The cookbook provides a new definition. At some point in the future this definition may be refactored into a lightweight resource and provider as suggested by [foodcritic rule FC015](http://acrmp.github.com/foodcritic/#FC015).

### nginx\_site

Enable or disable a Server Block in
`#{node['nginx']['dir']}/sites-available` by calling nxensite or
nxdissite (introduced by this cookbook) to manage the symbolic link in
`#{node['nginx']['dir']}/sites-enabled`.

### Parameters:

* `name` - Name of the site.
* `enable` - Default true, which uses `nxensite` to enable the site. If false, the site will be disabled with `nxdissite`.
* `template` - (optional) Path to the source for the `template` resource.
* `variables` - (optional) Variables to be used with the `template` resource


Ohai Plugin
-----------
The `ohai_plugin` recipe includes an Ohai plugin. It will be
automatically installed and activated, providing the following
attributes via ohai:

- `node['nginx']['version']` - version of nginx
- `node['nginx']['configure_arguments']` - options passed to
  `./configure` when nginx was built
- `node['nginx']['prefix']` - installation prefix
- `node['nginx']['conf_path']` - configuration file path

Usage
-----
Include the recipe on your node or role that fits how you wish to
install Nginx on your system per the recipes section above. Modify the
attributes as required in your role to change how various
configuration is applied per the attributes section above. In general,
override attributes in the role should be used when changing
attributes.

There's some redundancy in that the config handling hasn't been
separated from the installation method (yet), so use only one of the
recipes, or default.


License & Authors
-----------------
### Current Maintainers
- Author:: Eric Herot (<eric.herot@evertrue.com>)
- Author:: Jeff Byrnes (<jeff@evertrue.com>)

### Original Authors
- Author:: Joshua Timberman (<joshua@chef.io>)
- Author:: Adam Jacob (<adam@chef.io>)
- Author:: AJ Christensen (<aj@chef.io>)
- Author:: Jamie Winsor (<jamie@vialstudios.com>)
- Author:: Mike Fiedler (<miketheman@gmail.com>)

```text
Copyright 2014-2015, EverTrue, Inc and
Copyright 2008-2014, Chef Software, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
nginx Cookbook CHANGELOG
========================
This file is used to list changes made in each version of the nginx cookbook.

et-v6.1.1 (2019-02-06)
----------------------

* added $request_time to nginx log format

et-v6.1.0 (2017-08-06)
----------------------

* Switch to installing nginx-extras instead of nginx-full to get access to Lua extensions

et-v6.0.4 (2017-08-01)
----------------------

* Pin nginx package so that it is not auto-updated by "unattended-updates"
* Bump nginx package to latest for 16.04

et-v6.0.3 (2017-07-20)
----------------------

* Specify both nginx-full and nginx-common packages so that Evertrue-specific version numbers get used to install both

et-v6.0.2 (2017-05-31)
----------------------

* Grab different custom nginx packages on different versions of Ubuntu

et-v6.0.1 (2016-12-19)
----------------------

* Add original client IPs, precision timestamps, application keys, and authorization headers to nginx access logs

et-v6.0.0 (2016-09-22)
----------------------

* **BREAKING** Bump Ohai version to 4.0
* Stop checksumming GeoIP databases (seriously?)
* Merge branch '2.7.x' of github.com:miketheman/nginx into integration (not directly, but that's not important right now)
* Test exclusively on EC2

et-v5.1.0 (2016-04-27)
----------------------

* Make it possible for the default server to return a response code instead of a page
* Make it possible to configure the hostname for the default server

et-v5.0.0 (2015-09-28)
----------------------

* Prune a bunch of stuff that we don't use at EverTrue
    - Remove support for source install
    - Drop support for platforms that are not Debian/Ubuntu
    - Drop support for init styles other than Upstart
    - Drop passenger support
* Fix some minor errors in the ServerSpec tests
* Provide a default socketproxy app for testing
* Switch to custom build of the nginx package (called `nginx-full`) to get [upstream check module](https://github.com/yaoweibin/nginx_upstream_check_module) support
    - Use version when installing nginx package
* Remove spec test to verify nginx start (which we no longer do)
* Fix remaining incorrect namespacing (s/nginx/et_nginx/g) mostly in spec tests
* Remove orphaned object package_install_opts from package recipe
* Update the documentation to be more EverTrue specific
    - Do not attempt to document every single attribute in the metadata
* Update nginx Ohai plugin to new format
* Add socketproxy to modules test
* Renamed/moved recipes and templates to distinguish compile-time modules from other stuff
* Stop populating the `node['nginx']['version']` field with Ohai (it creates confusion)
* Move the package name to be installed to its own attributes file
* Modernize kitchen yaml
* Remove "which nginx" not_if from nginx package install
* Update the kitchen cloud ami (to: `ubuntu-trusty-14.04-amd64-server-20150810`)

et-v4.0.0 (2015-04-08)
----------------------

* Switch to using regular expressions for location matching on socketproxy (POTENTIALLY BREAKING)

et-v3.0.1 (2015-01-16)
----------------------

* Use compile-time "unless" instead of not_if to decide whether or not to make a symlink for the log directory

et-v3.0.0 (2015-01-16)
----------------------

* *BREAKING:* Use access_log_options attribute to set options for socketproxy access log (instead of assuming "main") - Fixes #308
* Add a policy-rc.d file on Debian systems to prevent the nginx service from starting when the package installs
* Lets only test on the system we actually use here at ET: amazon with ubuntu-1404
* Use et_nginx as the cookbook name in kitchen YAML instead of nginx
* Move log directory creation to before package install (and add user ourselves to support it instead of waiting for the package to add it)
* Stop deleting incorrect nginx log directory
* Serverspec tests should check for port 8080 (not 80)
* nginx user should use nologin binary instead of `/bin/false`

et-v2.0.6 (2014-10-06)
----------------------
-- Add actual Serverspec tests
-- Clean up Gemfile
-- Add additional guard to removal of `/var/log/nginx` to avoid failures due to `et_rails_app::default` installing a symlink at that path

v2.7.6 / 2015-03-17
==================

  * Bugfix sites do not need a .conf suffix anymore, [#338][] [@runningman84][]

v2.7.5 (2015-03-17)
-------------------
**NOTE** As of this release, this cookbook in its current format is deprecated,
and only critical bugs and fixes will be added.
A complete rewrite is in progress, so we appreciate your patience while we sort things out.
The amount of change included here

* Fix nginx 1.4.4 archive checksum to prevent redownload, [#305][] [@irontoby][]
* Allow setting an empty string to prevent additional repos, [#243][] [@miketheman][]
* Use correct `mime.types` for javascript, [#259][] [@dwradcliffe][]
* Fix `headers_more` module for source installs, [#279][], [@josh-padnick][] & [@miketheman][]
* Remove `libtool` from `geoip` and update download paths & checksums, [@miketheman][]
* Fix unquoted URL with params failing geoip module build (and tests!), [#294][] [@karsten-bruckmann][] & [@miketheman][]
* Fix typo in `source.rb`, [#205][] [@gregkare][]
* Test updates: ChefSpec, test-kitchen. Lots of help by [@jujugrrr][]
* Toolchain updates for testing
* Adds support for `tcp_nopush`, `tcp_nodelay` [@shtouff][]

After merging a ton of pull requests, here's a brief changelog. Click each to read more.

* Merge pull request [#335][] from [@stevenolen][]
* Merge pull request [#332][] from [@monsterstrike][]
* Merge pull request [#331][] from [@jalberto][]
* Merge pull request [#327][] from [@nkadel-skyhook][]
* Merge pull request [#326][] from [@bchrobot][]
* Merge pull request [#325][] from [@CanOfSpam3bug324][]
* Merge pull request [#321][] from [@jalberto][]
* Merge pull request [#318][] from [@evertrue][]
* Merge pull request [#314][] from [@bkw][]
* Merge pull request [#312][] from [@thomasmeeus][]
* Merge pull request [#310][] from [@morr][]

attributes/default.rb
#
# Cookbook Name:: nginx
# Attributes:: default
#
# Author:: Adam Jacob (<adam@chef.io>)
# Author:: Joshua Timberman (<joshua@chef.io>)
#
# Copyright 2009-2013, Chef Software, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# In order to update the version, the checksum attribute must be changed too.
# This attribute is in the source.rb file, though we recommend overriding
# attributes by modifying a role, or the node itself.
default['nginx']['version'] =
  value_for_platform(
    :ubuntu => {
      '14.04' => '1.4.6-2evertrue2.0',
      '16.04' => '1.10.3-1evertrue2.0'
    }
  )

default['nginx']['port']                = '80'
default['nginx']['dir']                 = '/etc/nginx'
default['nginx']['script_dir']          = '/usr/sbin'
default['nginx']['log_dir']             = '/var/log/nginx'
default['nginx']['log_dir_perm']        = '0750'
default['nginx']['binary']              = '/usr/sbin/nginx'
default['nginx']['default_hostname']    = node['hostname']
default['nginx']['default_return_code'] = nil
default['nginx']['default_root']        = '/var/www/nginx-default'
default['nginx']['ulimit']              = '1024'

default['nginx']['pid'] = '/var/run/nginx.pid'

default['nginx']['user'] = 'www-data'
if platform == 'ubuntu' && platform_version == '14.04'
  default['nginx']['pid'] = '/run/nginx.pid'
end

default['nginx']['upstart']['runlevels']     = '2345'
default['nginx']['upstart']['respawn_limit'] = nil
default['nginx']['upstart']['foreground']    = true

default['nginx']['group'] = node['nginx']['group'] || node['nginx']['user']

default['nginx']['gzip']              = 'on'
default['nginx']['gzip_static']       = 'off'
default['nginx']['gzip_http_version'] = '1.0'
default['nginx']['gzip_comp_level']   = '2'
default['nginx']['gzip_proxied']      = 'any'
default['nginx']['gzip_vary']         = 'off'
default['nginx']['gzip_buffers']      = nil
default['nginx']['gzip_types'] = %w(
  text/plain
  text/css
  application/x-javascript
  text/xml
  application/xml
  application/rss+xml
  application/atom+xml
  text/javascript
  application/javascript
  application/json
  text/mathml
)
default['nginx']['gzip_min_length']   = 1_000
default['nginx']['gzip_disable']      = 'MSIE [1-6]\.'

default['nginx']['keepalive']            = 'on'
default['nginx']['keepalive_requests']   = 100
default['nginx']['keepalive_timeout']    = 65
default['nginx']['worker_processes']     = node['cpu'] && node['cpu']['total'] ? node['cpu']['total'] : 1
default['nginx']['worker_connections']   = 1_024
default['nginx']['worker_rlimit_nofile'] = nil
default['nginx']['multi_accept']         = false
default['nginx']['event']                = nil
default['nginx']['accept_mutex_delay']   = nil
default['nginx']['server_tokens']        = nil
default['nginx']['server_names_hash_bucket_size'] = 64
default['nginx']['variables_hash_max_size']       = 1024
default['nginx']['variables_hash_bucket_size']    = 64
default['nginx']['sendfile'] = 'on'
default['nginx']['underscores_in_headers'] = nil
default['nginx']['tcp_nodelay'] = 'on'
default['nginx']['tcp_nopush'] = 'on'

default['nginx']['access_log_options']     = nil
default['nginx']['error_log_options']      = nil
default['nginx']['disable_access_log']     = false
default['nginx']['log_formats']            = {}
default['nginx']['default_site_enabled']   = true
default['nginx']['types_hash_max_size']    = 2_048
default['nginx']['types_hash_bucket_size'] = 64

default['nginx']['proxy_read_timeout']          = nil
default['nginx']['client_body_buffer_size']     = nil
default['nginx']['client_max_body_size']        = nil
default['nginx']['large_client_header_buffers'] = nil
default['nginx']['modules']                     = []

default['nginx']['extra_configs'] = {}
